File: web-admin/src/services/randomDiscount.ts
import request from '@/utils/request'

export interface DiscountSetting {
  configured: boolean
  id?: number
  event_id: number
  enabled: boolean
  min_discount_amount: number
  max_discount_amount: number
  probability: number
  total_pool: number
  remaining_pool: number
  max_discount_per_transaction: number | null
  min_payment_amount: number
  daily_limit_per_user: number | null
  created_at?: string
  updated_at?: string
}

export interface DiscountSettingRequest {
  event_id: number
  enabled: boolean
  min_discount_amount: number
  max_discount_amount: number
  probability: number
  total_pool: number
  max_discount_per_transaction?: number | null
  min_payment_amount: number
  daily_limit_per_user?: number | null
}

export interface DiscountStatistics {
  configured: boolean
  enabled: boolean
  total_pool: number
  remaining_pool: number
  used_pool: number
  total_discount_count: number
  total_discount_amount: number
  today_discount_count: number
  today_discount_amount: number
  min_discount_amount?: number
  max_discount_amount?: number
  probability?: number
  min_payment_amount?: number
  daily_limit_per_user?: number | null
}

export interface DiscountRecord {
  id: number
  participant_id: number
  transaction_id: number
  booth_id: number | null
  original_amount: number
  discount_amount: number
  actual_amount: number
  created_at: string
}

export interface DiscountRecordsResponse {
  records: DiscountRecord[]
  total_count: number
}

// 获取随机立减配置
export const getDiscountSetting = (eventId: number) => {
  return request.get<any, DiscountSetting>(`/random-discount/settings/${eventId}`)
}

// 创建或更新随机立减配置
export const saveDiscountSetting = (data: DiscountSettingRequest) => {
  return request.post<any, any>('/random-discount/settings', data)
}

// 重置奖池
export const resetDiscountPool = (eventId: number, newPool?: number) => {
  return request.post<any, any>(`/random-discount/settings/${eventId}/reset-pool`, {
    new_pool: newPool || null,
  })
}

// 获取统计信息
export const getDiscountStatistics = (eventId: number) => {
  return request.get<any, DiscountStatistics>(`/random-discount/statistics/${eventId}`)
}

// 获取立减记录
export const getDiscountRecords = (
  eventId: number,
  params?: { participant_id?: number; limit?: number; offset?: number }
) => {
  return request.get<any, DiscountRecordsResponse>(`/random-discount/records/${eventId}`, {
    params,
  })
}

File: models/random_discount.py
"""
Random Discount ORM models for NFC Campus E-Wallet System.

随机立减系统模型：配置表和记录表。
"""

from sqlalchemy import Column, Integer, String, DateTime, Boolean, ForeignKey, Numeric
from sqlalchemy.orm import relationship
from datetime import datetime, timezone

from core.database import Base


class RandomDiscountSetting(Base):
    """
    随机立减配置模型。
    
    每个活动可以有一个随机立减配置，管理员可以设置：
    - 立减金额范围（最小/最大）
    - 触发概率
    - 总奖池和剩余奖池
    - 单笔最大立减
    - 最低消费金额门槛
    - 每人每日限制次数
    """
    __tablename__ = 'random_discount_settings'
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    event_id = Column(Integer, ForeignKey('events.id', ondelete='CASCADE'), nullable=False, unique=True)
    enabled = Column(Boolean, nullable=False, default=False)
    min_discount_amount = Column(Numeric(12, 2), nullable=False, default=0.01)
    max_discount_amount = Column(Numeric(12, 2), nullable=False, default=5.00)
    probability = Column(Integer, nullable=False, default=100)  # 1-100
    total_pool = Column(Numeric(12, 2), nullable=False, default=1000.00)
    remaining_pool = Column(Numeric(12, 2), nullable=False, default=1000.00)
    max_discount_per_transaction = Column(Numeric(12, 2), nullable=True)
    min_payment_amount = Column(Numeric(12, 2), nullable=False, default=1.00)
    daily_limit_per_user = Column(Integer, nullable=True)
    created_at = Column(DateTime, nullable=False, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime, nullable=False, default=lambda: datetime.now(timezone.utc),
                        onupdate=lambda: datetime.now(timezone.utc))
    
    # Relationships
    event = relationship("Event", backref="random_discount_setting")
    records = relationship("RandomDiscountRecord", back_populates="setting",
                          foreign_keys="RandomDiscountRecord.event_id",
                          primaryjoin="RandomDiscountSetting.event_id==RandomDiscountRecord.event_id",
                          overlaps="event,random_discount_setting",
                          viewonly=True)


class RandomDiscountRecord(Base):
    """
    随机立减记录模型。
    
    记录每次随机立减的详细信息，用于审计和统计。
    """
    __tablename__ = 'random_discount_records'
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    event_id = Column(Integer, ForeignKey('events.id', ondelete='CASCADE'), nullable=False)
    participant_id = Column(Integer, ForeignKey('participants.id', ondelete='CASCADE'), nullable=False)
    transaction_id = Column(Integer, ForeignKey('transactions.id', ondelete='CASCADE'), nullable=False)
    booth_id = Column(Integer, ForeignKey('booths.id', ondelete='SET NULL'), nullable=True)
    original_amount = Column(Numeric(12, 2), nullable=False)
    discount_amount = Column(Numeric(12, 2), nullable=False)
    actual_amount = Column(Numeric(12, 2), nullable=False)
    created_at = Column(DateTime, nullable=False, default=lambda: datetime.now(timezone.utc))
    
    # Relationships
    event = relationship("Event", overlaps="random_discount_setting,records")
    participant = relationship("Participant")
    transaction = relationship("Transaction")
    booth = relationship("Booth")
    setting = relationship("RandomDiscountSetting",
                          foreign_keys=[event_id],
                          primaryjoin="RandomDiscountRecord.event_id==RandomDiscountSetting.event_id",
                          viewonly=True,
                          overlaps="event,records,random_discount_setting")
